libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,513 @@
+	for _, c := range cfg {
+		if !c.disableNAT {
+			addrs = append(addrs, c.HostIP)
+		}
 	}
 
-	// Save the host port (regardless it was or not specified in the binding)
-	switch netAddr := host.(type) {
-	case *net.TCPAddr:
-		bnd.HostPort = uint16(host.(*net.TCPAddr).Port)
-		return nil
-	case *net.UDPAddr:
-		bnd.HostPort = uint16(host.(*net.UDPAddr).Port)
-		return nil
-	case *sctp.SCTPAddr:
-		bnd.HostPort = uint16(host.(*sctp.SCTPAddr).Port)
-		return nil
-	default:
-		// For completeness
-		return ErrUnsupportedAddressType(fmt.Sprintf("%T", netAddr))
+	if len(addrs) > 0 {
+		pa := portallocator.Get()
+		port, err = pa.RequestPortsInRange(addrs, proto, int(hostPortStart), int(hostPortEnd))
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if retErr != nil {
+				for _, a := range addrs {
+					pa.ReleasePort(a, proto, port)
+				}
+			}
+		}()
+	}
+
+	res := make([]portBinding, 0, len(cfg))
+	for _, c := range cfg {
+		cc := c.PortBinding.GetCopy()
+		pb := portBinding{PortBinding: &cc}
+		if c.disableNAT {
+			pb.HostPort = 0
+		} else {
+			pb.proxy, pb.stopProxy, err = startProxy(c.Proto.String(), c.HostIP, port, c.IP, int(c.Port), proxyPath)
+			if err != nil {
+				return nil, fmt.Errorf("failed to bind port %s:%d/%s: %w", c.HostIP, port, c.Proto, err)
+			}
+			defer func() {
+				if retErr != nil {
+					pb.stopProxy()
+				}
+			}()
+			pb.HostPort = uint16(port)
+		}
+		pb.HostPortEnd = pb.HostPort
+		res = append(res, pb)
 	}
+	return res, nil
 }
 
+// releasePorts attempts to release all port bindings, does not stop on failure
 func (n *bridgeNetwork) releasePorts(ep *bridgeEndpoint) error {
-	return n.releasePortsInternal(ep.portMapping)
-}
+	n.Lock()
+	pbs := ep.portMapping
+	ep.portMapping = nil
+	n.Unlock()
 
-func (n *bridgeNetwork) releasePortsInternal(bindings []types.PortBinding) error {
-	var errorBuf bytes.Buffer
+	return n.releasePortBindings(pbs)
+}
 
-	// Attempt to release all port bindings, do not stop on failure
-	for _, m := range bindings {
-		if err := n.releasePort(m); err != nil {
-			errorBuf.WriteString(fmt.Sprintf("\ncould not release %v because of %v", m, err))
+func (n *bridgeNetwork) releasePortBindings(pbs []portBinding) error {
+	var errs []error
+	for _, pb := range pbs {
+		var errP error
+		if pb.stopProxy != nil {
+			errP = pb.stopProxy()
+			if errP != nil {
+				errP = fmt.Errorf("failed to stop docker-proxy for port mapping %s: %w", pb, errP)
+			}
 		}
+		errN := n.setPerPortIptables(pb, false)
+		if errN != nil {
+			errN = fmt.Errorf("failed to remove iptables rules for port mapping %s: %w", pb, errN)
+		}
+		if pb.HostPort > 0 {
+			portallocator.Get().ReleasePort(pb.HostIP, pb.Proto.String(), int(pb.HostPort))
+		}
+		errs = append(errs, errP, errN)
+	}
+	return errors.Join(errs...)
+}
+
+func (n *bridgeNetwork) setPerPortIptables(b portBinding, enable bool) error {
+	if len(b.IP) != len(b.HostIP) { // IPv4 addresses have been unmapped.
```

Comment:
Don't assume that IPv4 addresses have been unmapped. An IPv4 address is an IPv4 address is an IPv4 address once it hits the wire, whether it is encoded in four bytes or in sixteen.
```suggestion
	if (b.IP.To4() != nil) != (b.HostIP.To4() != nil) {
```

Reply:
Oh, cool - just did that, then saw this comment.